addons/payment_paypal/models/paypal_s2s.py

```diff
@@ -0,0 +1,231 @@
+                res = e.read()
+                e.close()
+                if tries and res and json.loads(res)['name'] == 'INTERNAL_SERVICE_ERROR':
+                    _logger.warning('Failed contacting Paypal, retrying (%s remaining)' % tries)
+            tries = tries - 1
+        if not res:
+            pass
+        result = json.loads(res.read())
+        res.close()
+        return result
+
+    def _paypal_s2s_send(self):
+        path = '/v1/payments/payment'
+        payment_url = self.env['payment.acquirer']._get_paypal_urls(self.acquirer_id.environment)['paypal_rest_url'] + path
+
+        amount_total = self.amount
+        amount = "%0.2f" % float(amount_total)
+
+        access_token = self.acquirer_id._paypal_s2s_get_access_token()
+        payment_headers = {
+            'Authorization': 'Bearer %s' % access_token,
+            'Content-Type': 'application/json'
+        }
+        data = {
+            'intent': 'sale',
+            'payer': {
+                'payment_method': 'credit_card',
+                'funding_instruments': [{
+                    'credit_card_token': {
+                        'credit_card_id': str(self.payment_method_id.acquirer_ref),
+                        'payer_id': str(self.payment_method_id.partner_id.id)
+                    }
+                }]
+            },
+            'transactions': [{
+                'amount': {
+                    'total': str(amount),
+                    'currency': str(self.currency_id.name),
+                },
+                'description': str(self.display_name)
+            }]
+        }
+        payment_result = dict()
+        try:
+            payment_request = urllib2.Request(payment_url,
+                                              json.dumps(data),
+                                              payment_headers)
+            payment_result = self._paypal_try_url(payment_request, tries=3)
+
+        except Exception, e:
+            _logger.exception('PayPal create payment request failed - %s' % e)
+            raise UserWarning(_('PayPal create payment request failed.'))
+        return payment_result
+
+    @api.multi
+    def paypal_s2s_do_transaction(self, **vals):
+        self.ensure_one()
+        payment_method = self.env['payment.method'].search([
+            ('partner_id', '=', self.partner_id.id),
+            ('acquirer_id', '=', self.acquirer_id.id)])
+        if not payment_method:
+            raise UserWarning(_('Credit/Debit card has been not added in your account.'))
+
+        payment_result = self._paypal_s2s_send()
+
+        return self._paypal_s2s_validate_tree(payment_result)
+
+    @api.cr_uid_records_context
+    def paypal_s2s_form_validate(self, cr, uid, tx, context=None):
+        tree = self._paypal_s2s_get_tx_status(tx)
+        return self._paypal_s2s_validate_tree(tx, tree)
+
+    @api.multi
+    def _paypal_s2s_validate_tree(self, data):
+        response = data['state']
```

Comment:
First check if the tx was already validated (if the tx is not in draft or pending state, it was already validated and the rest of this code should not be executed)